src/sidebar/components/markdown-editor.js

```diff
@@ -129,74 +133,213 @@ function ToolbarButton({
 }
 
 ToolbarButton.propTypes = {
+  buttonRef: propTypes.object.isRequired,
   disabled: propTypes.bool,
   iconName: propTypes.string,
   label: propTypes.string,
-  onClick: propTypes.func,
+  onClick: propTypes.func.isRequired,
   shortcutKey: propTypes.string,
+  tabIndex: propTypes.number.isRequired,
   title: propTypes.string,
 };
 
+/**
+ * @typedef {string} ButtonID
+ * A unique string that can be one of one of:
+ * [bold, italic, quote, link, image, math, numlist, list, preview, help]
+ */
+
+/**
+ * An array of toolbar elements with a roving tab stop. Left and right
+ * array keys can be used to change focus of the elements. Home and end
+ * keys will navigate to the first and last elements respectively.
+ *
+ * Canonical example
+ * https://www.w3.org/TR/wai-aria-practices/examples/toolbar/toolbar.html
+ */
 function Toolbar({ isPreviewing, onCommand, onTogglePreview }) {
+  const buttonIds = {
+    // Ordered buttons
+    bold: 0,
+    italic: 1,
+    quote: 2,
+    link: 3,
+    image: 4,
+    math: 5,
+    numlist: 6,
+    list: 7,
+    help: 8,
+    preview: 9,
+
+    // Total button count
+    maxId: 10,
+  };
+
+  // Keep track of a roving index. The active roving tabIndex
+  // is set to 0, and all other elements are set to -1.
+  const [rovingElement, setRovingElement] = useState(0);
+
+  // An array of refs
+  const buttonRefs = useRef([]).current;
+  if (buttonRefs.length === 0) {
+    // Initialize buttonRefs on first render only
+    for (let i = 0; i <= buttonIds.maxId; i++) {
+      buttonRefs.push(createRef());
+    }
+  }
+
+  /**
+   * Returns the unique button ref associated with the id.
+   *
+   * @param {ButtonID} id
+   * @return {Ref}
+   */
+  const getButtonRef = id => {
```

Comment:
Ah, I was thinking that you would pass the index value rather than the ID directly, ie. using `getButtonRef(buttonIds.button)`. In other words, the `buttonIds` object serves purely to group a bunch of number constants together.

One advantage of such a change is that it would be easier for a developer or static analysis tool to verify that `buttonIds.<some property name>` is a valid property name.

Reply:
right, if the keys change, its a syntax error rather than a less obvious logical error.